StortSpelProjekt/Engine/src/Renderer/TextManager.cpp

```diff
@@ -0,0 +1,263 @@
+#include "stdafx.h"
+#include "TextManager.h"
+
+#include "../Renderer/Renderer.h"
+#include "../Renderer/Texture/Texture.h"
+#include "../Renderer/DescriptorHeap.h"
+
+#include "../Misc/Window.h"
+#include "../Misc/AssetLoader.h"
+#include "../Misc/GUI2DElements/Text.h"
+#include "../Misc/GUI2DElements/Font.h"
+
+TextManager::TextManager()
+{
+	// Default text
+	m_DefaultTextData.color = { 1.0f, 1.0f, 1.0f, 1.0f };
+	m_DefaultTextData.padding = { 0.5f, 0.0f };
+	m_DefaultTextData.pos = { 0.5f, 0.5f };
+	m_DefaultTextData.scale = { 1.0f, 1.0f };
+	m_DefaultTextData.text = L"DEFAULT_TEXT";
+}
+
+TextManager::~TextManager()
+{
+	for (auto textMap : m_TextMap)
+	{
+		if (textMap.second != nullptr)
+		{
+			delete textMap.second;
+		}
+	}
+	m_TextMap.clear();
+}
+
+std::map<std::string, TextData>* const TextManager::GetTextDataMap()
+{
+	return &m_TextDataMap;
+}
+
+TextData* TextManager::GetTextData(std::string name)
+{
+	return &m_TextDataMap[name];
+}
+
+void TextManager::AddText(std::string name)
+{
+	// Look if a font is chosen, otherwise, set the default font
+	if (m_pFont == nullptr)
+	{
+		m_pFont = AssetLoader::Get()->LoadFontFromFile(L"Arial.fnt");
+	}
+
+	// Default text
+	TextData textData = m_DefaultTextData;
+
+	auto it = m_TextDataMap.find(name);
+	if (it != m_TextDataMap.end())
+	{
+		Log::PrintSeverity(Log::Severity::WARNING, "It already exists a text with the name '%s'! Overwriting text data...\n", name.c_str());
+	}
+
+	m_TextDataMap[name] = textData;
+}
+
+void TextManager::UploadTextData(std::string name)
+{
+	Renderer* renderer = &Renderer::GetInstance();
+
+	int numOfCharacters = GetNumOfCharacters(name);
+	auto textData = GetTextData(name);
+
+	Text* text = new Text(
+		renderer->m_pDevice5,
+		renderer->m_DescriptorHeaps[DESCRIPTOR_HEAP_TYPE::CBV_UAV_SRV],
+		numOfCharacters,
+		m_pFont->GetTexture());
+	text->SetTextData(textData, m_pFont);
+
+	// Look if the text exists
+	bool exists = false;
+	auto it = m_TextDataMap.find(name);
+	if (it != m_TextDataMap.end())
+	{
+		exists = true;
+	}
+
+	if (exists == true)
+	{
+		// Replacing an existing text in the map
+		replaceText(text, name);
+	}
+	else
+	{
+		// Adding a new text to the map
+		submitText(text, name);
+	}
+
+	// Uploading the text data to the gpu
+	renderer->submitTextToGPU(text, this);
+
+	renderer->executeCopyOnDemand();
```

Comment:
Will this be executed in the middle of a "initScene" now aswell? Lets say we are adding 20 orcs, first 10 orcs, all submitting to the copyTask.. and then you add a text. Will this code execute the command lists at that moment? Because we would rather want to wait before everything is recorded in the commandlists before we execute.